From 4aebf3f99c6e546a0cadb500891ba1a34640f295 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Mon, 9 Sep 2024 11:13:01 +0200
Subject: [PATCH 680/841] refactor(stm32mp2): update authentication check for
 STM32MP2
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The platform allow authentication check when chip is open to comfirm
that the key hash stored in OTP is correct. Put all that code directly
in authentication_check() function in order to simplify it.

In the STM32 MPU, we use auth_status from boot_api_context to display
the authentication check status on the console for CLOSED
UNSECURED/UNLOCKED samples. Similar to the MP25, the MP21 status
reflects only the signature check and not the ROTPK check.

However, unlike the STM32MP25, the STM32MP21 platform doesn't allow
reading the OTP that contains the ROTPK Hash, so there is no way to be
sure that the authentication succeeded when the chip is in an open
state. The authentication check code is then useless in the STM32MP21,
so the dedicated portion of code should be stubbed.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git fb26238191bbf68ebaa6ce980f08db60d56cae47 refactor(stm32mp2): update authentication check for STM32MP2]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: Id20e813cdff751d000011fc33a55c90065076966
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/403011
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 35 +++++++++++++++++--------------
 1 file changed, 19 insertions(+), 16 deletions(-)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index ae1f2c30f..fac62a5a6 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -337,11 +337,13 @@ static void check_tamper_event(bool lse_tamper_occured)
 	}
 }
 
-#if TRUSTED_BOARD_BOOT && DYN_DISABLE_AUTH
-static bool authentication_check(boot_api_context_t *boot_context)
+static void authentication_check(boot_api_context_t *boot_context)
 {
+#if TRUSTED_BOARD_BOOT && DYN_DISABLE_AUTH && !STM32MP21
+	bool auth = false;
+
 	if (boot_context->auth_status == BOOT_API_CTX_AUTH_FAILED) {
-		return false;
+		goto end;
 	}
 
 	if (stm32mp_check_closed_device() != STM32MP_CHIP_SEC_CLOSED) {
@@ -350,15 +352,24 @@ static bool authentication_check(boot_api_context_t *boot_context)
 		unsigned int flags = 0U;
 		int rc;
 
+		if (stm32_hash_register() != 0) {
+			ERROR("%s: HASH register fail\n", __func__);
+			return;
+		}
+
 		rc = plat_get_rotpk_info(NULL, &pk_ptr, &len, &flags);
-		if ((rc == -EINVAL) || ((flags & ROTPK_NOT_DEPLOYED) > 0)) {
-			return false;
+		if ((rc != -EINVAL) && ((flags & ROTPK_NOT_DEPLOYED) == 0)) {
+			auth = true;
 		}
+	} else {
+		auth = true;
 	}
+end:
+	NOTICE("Bootrom authentication %s\n", auth ? "succeeded" : "failed");
 
-	return true;
+#endif /* TRUSTED_BOARD_BOOT && DYN_DISABLE_AUTH && !STM32MP21 */
 }
-#endif /* TRUSTED_BOARD_BOOT && DYN_DISABLE_AUTH */
+
 
 void bl2_el3_plat_arch_setup(void)
 {
@@ -455,17 +466,9 @@ void bl2_el3_plat_arch_setup(void)
 
 	print_reset_reason();
 
-#if TRUSTED_BOARD_BOOT && DYN_DISABLE_AUTH
-	if (stm32_hash_register() != 0) {
-		ERROR("HASH register fail\n");
-		panic();
-	}
-
 	if (boot_context->auth_status != BOOT_API_CTX_AUTH_NO) {
-		NOTICE("Bootrom authentication %s\n",
-		       authentication_check(boot_context) ? "succeeded" : "failed");
+		authentication_check(boot_context);
 	}
-#endif /* TRUSTED_BOARD_BOOT && DYN_DISABLE_AUTH */
 
 skip_console_init:
 	check_tamper_event(lse_tamper_occured);
-- 
2.47.2

